Add `label_selector` option to remote functions

ryanaoleary opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Adds the `label_selector` option to Ray tasks/actors through `_common_options`, to specify labels used for scheduling on Ray nodes.

## Related issue number

https://github.com/ray-project/ray/issues/51564

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

MengjinYan commented on python/ray/actor.py:
```diff
@@ -538,6 +538,9 @@ class _ActorClassMetadata:
             task.
         memory: The heap memory quota for this actor.
         resources: The default resources required by the actor creation task.
+        label_selector: The labels required for the node on which this actor
+            runs. Label selectors consist of a key-value pair, where the key
+            is a label name and the value is a label value with a condition.
```
```suggestion
        label_selector: The labels required for the node on which this actor
            can be scheduled on. The label selector consist of key-value pairs, where the keys
            are label names and the value are expressions consisting of an operator with label values or just a value to indicate equality..
```

I think it will also be better to add the supported operators and some examples of the label expressions

MengjinYan commented on python/ray/remote_function.py:
```diff
@@ -57,6 +57,8 @@ class RemoteFunction:
             remote function.
         _memory: The heap memory request in bytes for this task/actor,
             rounded down to the nearest integer.
+        _label_selector: The label requirements on a node for invocations of
+            this remote function.
```
```suggestion
        _label_selector: The label requirements on a node for scheduling of the task or actor.
```

MengjinYan commented on python/ray/actor.py:
```diff
@@ -777,6 +782,8 @@ def options(self, **actor_options):
             resources (Dict[str, float]): The quantity of various custom resources
                 to reserve for this task or for the lifetime of the actor.
                 This is a dictionary mapping strings (resource names) to floats.
+            label_selector (Dict[str, str]): If specified, requires that the task or actor run
```
```suggestion
            label_selector (Dict[str, str]): If specified, requires that the actor run
```

MengjinYan commented on python/ray/tests/test_actor.py:
```diff
@@ -691,6 +691,12 @@ class Actor:  # noqa: F811
+    class Actor:  # noqa: F811
+        def __init__(self):
+            pass
+
```
Can you add tests for (1) all the supported operators (2) with multiple labels and (3) some invalid case as well?

ryanaoleary replied:
Done in [0b09cd5](https://github.com/ray-project/ray/pull/51707/commits/0b09cd5c6d10c6e66be57dc2fa41316534668708).

MengjinYan reviewed (commented):
Thanks for the effort! 

I have one general question: It seems that this PR only contains the logic to add the options. For the validation logic, are you planning to add it in a future PR?

ryanaoleary commented:
> Thanks for the effort!
> 
> I have one general question: It seems that this PR only contains the logic to add the options. For the validation logic, are you planning to add it in a future PR?

It already validates that `label_selector` is a valid remote function option [here](https://github.com/ray-project/ray/blob/5195be51f484900f1ab385906a8592369032dbe2/python/ray/actor.py#L869). For other Actor options, I don't see additional validation about what's passed in as an argument besides that check. Should we be validating here that the labels passed in follow the expected syntax (the same as we do for the node labels) and that the conditions are supported? Or is it sufficient just to allow the user to pass in the `label_selector` and only schedule it to the node if they are valid.

MengjinYan commented:
> > Thanks for the effort!
> > I have one general question: It seems that this PR only contains the logic to add the options. For the validation logic, are you planning to add it in a future PR?
> 
> It already validates that `label_selector` is a valid remote function option [here](https://github.com/ray-project/ray/blob/5195be51f484900f1ab385906a8592369032dbe2/python/ray/actor.py#L869). For other Actor options, I don't see additional validation about what's passed in as an argument besides that check. Should we be validating here that the labels passed in follow the expected syntax (the same as we do for the node labels) and that the conditions are supported? Or is it sufficient just to allow the user to pass in the `label_selector` and only schedule it to the node if they are valid.

I think we definitely shouldn't silently ignore the selectors if they are not valid. We need to give users explicitly warnings to users. Personally, I would prefer we check the syntax here and raise an error early so that the users can be aware of the invalid syntax and fix accordingly.

MengjinYan commented on python/ray/_private/utils.py:
```diff
@@ -82,6 +82,21 @@
 
+# Regex patterns used to validate that labels conform to Kubernetes label syntax rules.
+# https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set
+# Regex for mandatory name (DNS label) or value
```
Can you add some examples in the comments about what each regex does for people to understand better?

ryanaoleary replied:
Done in [4878652](https://github.com/ray-project/ray/pull/51707/commits/48786520fab61449c8b273708a277897b1d03796)

MengjinYan commented on python/ray/_private/utils.py:
```diff
@@ -1831,13 +1846,65 @@ def parse_node_labels_json(
     if labels is None:
         return
-    for key in labels.keys():
+    for key, value in labels.keys():
```
I'm wondering how this logic will work with the default label passed in by kuberay. 

My understanding is when kuberay passes in the default labels, it will also use the `--labels` option in the `ray start` command and the set of labels will be validated here. If that's the case, then it is possible for ray reserved labels to be validated here. And this validation will cause false positives.

ryanaoleary replied:
Yeah I think so, we could remove the check for `ray.io/` prefixes when we add the KubeRay logic and put the onus on the user not to define Ray labels themselves. I'm thinking that even in the case where a user manually sets a `ray.io` label, we could overwrite it if it's one of the default labels that it's necessary for Ray to set. What do you think?

MengjinYan replied:
I think the logic should be added on KubeRay/autoscaler side to make sure the user specified labels cannot be Ray default labels, because that's the place we have the differentiation. And we can remove the check here.

ryanaoleary replied:
Sounds good to me, removed the check in [47d4794](https://github.com/ray-project/ray/pull/51707/commits/47d4794f5b1ea8aad9d7f66c175d4422b3cabfe3).

MengjinYan commented on python/ray/_private/utils.py:
```diff
@@ -1831,13 +1846,65 @@ def parse_node_labels_json(
+        return None
+    if not re.match(LABEL_SELECTOR_REGEX, selector):
+        return str(
+            f"Invalid label selector value `{selector}`. Supported operators are: ! and {LABEL_OPERATORS}. "
```
```suggestion
            f"Invalid label selector value `{selector}`. The label selector value should contain optional operators and a label value. Supported operators are: ! and {LABEL_OPERATORS}. "
```

MengjinYan reviewed (approved):
Thanks!

edoakes commented on python/ray/_private/utils.py:
```diff
@@ -82,6 +82,34 @@
 # Match the standard alphabet used for UUIDs.
 RANDOM_STRING_ALPHABET = string.ascii_lowercase + string.digits
 
+# Regex patterns used to validate that labels conform to Kubernetes label syntax rules.
+# https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set
+
+# Regex for mandatory name (DNS label) or value
```
same as other PR, let's move these to `label_utils.py`

ryanaoleary replied:
Sounds good, once https://github.com/ray-project/ray/pull/51706 is merged I'll pull in the changes from that PR.

ryanaoleary replied:
[f16a18d](https://github.com/ray-project/ray/pull/51707/commits/f16a18df34c00e081389e7f738d0c46573f2edb2)

edoakes commented on python/ray/_private/utils.py:
```diff
@@ -1831,13 +1859,59 @@ def parse_node_labels_json(
-                f"This is reserved for Ray defined labels."
+    for key, value in labels.items():
+        possible_error_message = validate_label_key(key)
+        if possible_error_message:
+            raise ValueError(possible_error_message)
+        if value is not None:
+            possible_error_message = validate_label_value(value)
+            if possible_error_message:
+                raise ValueError(possible_error_message)
```
hm this pattern is pretty strange, why not have `validate_label_value` directly raise the exception?

ryanaoleary replied:
I was following the same pattern as this function: https://github.com/ray-project/ray/blob/5b37f1cfbfaa519b251144ec1226e1d3cd5335ed/python/ray/_private/ray_option_utils.py#L104, it returns an `Optional[str]` because it gets set to this value here when the options are built.

ryanaoleary replied:
Just realized that this only needs to be the case for `validate_label_selector_value` and `validate_label_key` since that's what gets called in the ray option utils. I'll change `validate_label-value` to directly raise the exception.

ryanaoleary replied:
[4e0c3b0](https://github.com/ray-project/ray/pull/51707/commits/4e0c3b0cff03079ad910287ef9486bc3340bf94a)

edoakes commented on python/ray/actor.py:
```diff
@@ -538,6 +538,9 @@ class _ActorClassMetadata:
             task.
         memory: The heap memory quota for this actor.
         resources: The default resources required by the actor creation task.
+        label_selector: The labels required for the node on which this actor
+            can be scheduled on. The label selector consist of key-value pairs, where the keys
+            are label names and the value are expressions consisting of an operator with label values or just a value to indicate equality.
```
(just musing): this is where having a class-based API is nice for documentation purposes. you can link to the docstring of that thing

we should make sure to link to the documentation page for labels here once it exists

edoakes replied:
please also have the verbiage here directly match that in the `@ray.remote` decorator docstring

ryanaoleary replied:
Sounds good, done in [ddc024e](https://github.com/ray-project/ray/pull/51707/commits/ddc024e59d949325c61f2915a60a0037f9c67ff4)

ryanaoleary commented on python/ray/tests/test_actor.py:
```diff
@@ -692,6 +692,102 @@ def __init__(self):
             pass
 
 
+def test_decorator_label_selector_args(ray_start_regular_shared):
```
[101a99e](https://github.com/ray-project/ray/pull/51707/commits/101a99e7409e32352525eefad8da2df45d0b03e2)

ryanaoleary commented:
@edoakes I'm working on adding the logic to pass the `label_selector` that's passed in from the actor/task options to the core worker and then to the `TaskSpec` in this PR now. We'd then use that value in this PR: https://github.com/ray-project/ray/pull/51901 to do the actual scheduling. Once these two changes are merged, I can add the change in KubeRay to pass the default labels (i.e. market-type) to the `rayStartParams` and then I think all the basic pieces for the demo are finished.

edoakes commented:
@ryanaoleary there are some merge conflicts due to the revert, pls fix one the other PR is re-merged

ryanaoleary commented:
> @ryanaoleary there are some merge conflicts due to the revert, pls fix one the other PR is re-merged

@edoakes I've fixed the merge conflicts now that the other PR is merged, I also updated the validation logic to retain the previous check based on the discussion from the other PR.

edoakes commented:
> > @ryanaoleary there are some merge conflicts due to the revert, pls fix one the other PR is re-merged
> 
> @edoakes I've fixed the merge conflicts now that the other PR is merged, I also updated the validation logic to retain the previous check based on the discussion from the other PR.

There's another merge conflict 😓

ryanaoleary commented:
> > > @ryanaoleary there are some merge conflicts due to the revert, pls fix one the other PR is re-merged
> > 
> > 
> > @edoakes I've fixed the merge conflicts now that the other PR is merged, I also updated the validation logic to retain the previous check based on the discussion from the other PR.
> 
> There's another merge conflict 😓

Ah lol should be fixed now

edoakes commented on python/ray/_private/worker.py:
```diff
@@ -3553,6 +3554,10 @@ def method(self):
             to reserve for this task or for the lifetime of the actor.
             This is a dictionary mapping strings (resource names) to floats.
             By default it is empty.
+        label_selector (Dict[str, str]): If specified, the labels required for the node on
+                which this actor can be scheduled on. The label selector consist of key-value pairs,
+                where the keys are label names and the value are expressions consisting of an operator
+                with label values or just a value to indicate equality.
```
let's mark this as experimental in public-facing docstrings for now

ryanaoleary replied:
Fixed in 4b70316

edoakes reviewed (approved):
Looks good, minor nit

edoakes commented on python/ray/tests/test_node_labels.py:
```diff
@@ -155,18 +143,21 @@ def test_autoscaler_set_node_labels(autoscaler_v2, shutdown_only):
         assert node_info["Labels"] == add_default_labels(
             node_info, {"gpu_type": "A100", "region": "us", "market-type": "spot"}
         )
+    finally:
         subprocess.check_call(["ray", "stop", "--force"])
+        os.remove(test_file_path)
```
will merge the PR as-is, but please follow up to move all such cleanup logic in this file into fixtures as best practice
